### DISCUSSION TITLE
Build fails when `exactOptionalPropertyTypes` TS config option is enabled

### DISCUSSION THREAD

**Discussion by ziyadedher** [vercel/next.js#33712] [Ideas]
### Run `next info` (available from version 12.0.8 and up)

```
Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP PREEMPT Wed, 22 Dec 2021 09:23:54 +0000
    Binaries:
      Node: 17.3.0
      npm: N/A
      Yarn: 3.1.1
      pnpm: N/A
    Relevant packages:
      next: 12.0.8
      react: 17.0.2
      react-dom: 17.0.2
```

### What version of Next.js are you using?

12.0.8

### What version of Node.js are you using?

17.3.0

### What browser are you using?

Chrome (but irrelevant)

### What operating system are you using?

Arch Linux

### How are you deploying your application?

next build

### Describe the Bug

When the TypeScript `compilerOptions` option `exactOptionalPropertyTypes` is enabled, `next build` fails with the following error:
```
Failed to compile.

../../node_modules/next/dist/server/dev/next-dev-server.d.ts:63:5
Type error: Property 'generateRoutes' in type 'DevServer' is not assignable to the same property in base type 'NextNodeServer'.
```

My understanding is that `DevServer` has an "explicitly" optional type containing `{generateRoutes: Route | undefined}` while `NextNodeServer` has an "implicit" optional type containing `{generateRoutes?: Route}`. `exactOptionalPropertyTypes` disallows explicitly assigning `undefined` to optional types.

### Expected Behavior

I'm unsure if I _expect_ this to pass. But it would probably be good to be compatible with this option.

This bug report might be more of an improvement suggestion.

### To Reproduce

1. Set up a Typescript Next.js project normally (i.e. via the example or otherwise).
2. Modify `tsconfig.json` to have the following (along with what was there to begin with):
```json
{
  "compilerOptions": {
    "exactOptionalPropertyTypes": false,
  }
}
```
3. Run `next build` (`next dev` works fine), and watch it fail on the "Checking validity of types" step.

**Comment by balazsorban44** [comment]
Hi, I could not reproduce this using:

`yarn create next-app --ts`
`yarn build`

Works fine. Could you please add a more elaborate reproduction?

**Comment by ziyadedher** [comment]
Ah sorry, you also need to set `skipLibCheck: false` instead of `skipLibCheck: true` which is in the default configuration.

Here is a full reproduction:
1. `yarn create next-app --ts`
2. Edit `tsconfig.json` such that
    - `compilerOptions.skipLibCheck` is `false`.
    - `compilerOptions.exactOptionalPropertyTypes` is `true`
3. `yarn build`

![image](https://user-images.githubusercontent.com/16349988/151233485-d9f79c04-f5be-4e6a-9a35-6beec40f4330.png)
![image](https://user-images.githubusercontent.com/16349988/151233575-f03bbbec-2cc1-4d20-94eb-08f34b51ff16.png)

**Comment by balazsorban44** [comment]
In this case, it seems like a feature request. I'm converting this to a discussion. Feel free to open a PR to fix this.
